os/drivers/seclink/seclink_drv.c

```diff
@@ -0,0 +1,202 @@
+ * Included Files
+ ****************************************************************************/
+#include <tinyara/config.h>
+
+#include <unistd.h>
+#include <time.h>
+#include <errno.h>
+#include <string.h>
+#include <stdlib.h>
+#include <debug.h>
+
+#include <tinyara/fs/fs.h>
+#include <tinyara/testcase_drv.h>
+#include <tinyara/sched.h>
+#include <tinyara/seclink.h>
+#include <tinyara/seclink_drv.h>
+#include "seclink_drv_req.h"
+#include "seclink_drv_utils.h"
+
+
+/****************************************************************************
+ * Private Function Prototypes
+ ****************************************************************************/
+
+static int seclink_open(FAR struct file *filep);
+static int seclink_close(FAR struct file *filep);
+static ssize_t seclink_read(FAR struct file *filep, FAR char *buffer, size_t len);
+static ssize_t seclink_write(FAR struct file *filep, FAR const char *buffer, size_t len);
+static int seclink_ioctl(FAR struct file *filep, int cmd, unsigned long arg);
+
+/****************************************************************************
+ * Private Data
+ ****************************************************************************/
+
+static const struct file_operations g_seclink_fops = {
+	seclink_open,                                                   /* open */
+	seclink_close,                                                   /* close */
+	seclink_read,
+	seclink_write,
+	0,                                                   /* seek */
+	seclink_ioctl                                        /* ioctl */
+#ifndef CONFIG_DISABLE_POLL
+	, 0                                                  /* poll */
+#endif
+};
+
+int seclink_open(FAR struct file *filep)
+{
+	SLDRV_ENTER;
+
+	FAR struct inode *inode = filep->f_inode;
+	FAR struct sec_upperhalf_s *upper = inode->i_private;
+	upper->refcnt++;
+
+	return 0;
+}
+
+int seclink_close(FAR struct file *filep)
+{
+	SLDRV_ENTER;
+
+	FAR struct inode *inode = filep->f_inode;
+	FAR struct sec_upperhalf_s *upper = inode->i_private;
+	upper->refcnt--;
+
+	return 0;
+}
+
+ssize_t seclink_read(FAR struct file *filep, FAR char *buffer, size_t len)
+{
+	SLDRV_ENTER;
+
+	return 0;
+}
+
+ssize_t seclink_write(FAR struct file *filep, FAR const char *buffer, size_t len)
+{
+	SLDRV_ENTER;
+
+	return 0;
+}
+
+int seclink_ioctl(FAR struct file *filep, int cmd, unsigned long arg)
+{
+	SLDRV_ENTER;
+
+	SLDRV_LOG("-->%s (%d)(%x)\n", __FUNCTION__, cmd, arg);
+
+	FAR struct inode *inode = filep->f_inode;
+	FAR struct sec_upperhalf_s *upper = inode->i_private;
+
+	/* If there are a couple of tasks that request SE then
+	 * calling lower's function should be protected. But
+	 * current design consider that only one task that is seclink can
+	 * request to HAL
+	 */
+	if (cmd == SECLINK_HAL_INIT || cmd == SECLINK_HAL_DEINIT) {
+		hd_handle_common_request(cmd, arg, (void *)upper->lower);
+	} else if (cmd & SECLINK_AUTH) {
+		hd_handle_auth_reqeust(cmd, arg, (void *)upper->lower);
```

Comment:
request *

Reply:
sorry I can't understand what do you want to. please tell me more detail.